### PR TITLE
Potential fix for stress test failure due to "SST file ahead of WAL" error

### PR DESCRIPTION
**Pull request by anand1976**
I'm not able to prove it, but the stress test failure may be caused by the following sequence of events -

1. Crash db_stress while writing the log file. This should result in a corrupted WAL.
2. Run db_stress with recycle_log_file_num=1. Crash during recovery immediately after writing manifest and updating the current file. The old log from the previous run is left behind, but the memtable would have been flushed during recovery and the CF log number will point to the newer log
3. Run db_stress with recycle_log_file_num=0. During recovery, the old log file will be processed and the corruption will be detected. Since the CF has moved ahead, we get the "SST file is ahead of WAL" error

Test -
1. stress_crash
2. make check

### CODE REVIEW

**Comment by siying** [comment]
I can't see any problem with the PR, but I would let @riversand963 to decide.

File: db/db_impl/db_impl_open.cc [hunk 1]
```diff
@@ -588,12 +588,13 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
...
   uint64_t corrupted_log_number = kMaxSequenceNumber;
+  uint64_t min_log_number = MinLogNumberToKeep();
   for (auto log_number : log_numbers) {
-    if (log_number < versions_->min_log_number_to_keep_2pc()) {
+    if (log_number < min_log_number) {
       ROCKS_LOG_INFO(immutable_db_options_.info_log,
                      "Skipping log #%" PRIu64
                      " since it is older than min log to keep #%" PRIu64,
-                     log_number, versions_->min_log_number_to_keep_2pc());
+                     log_number, MinLogNumberToKeep());
```
Comment (riversand963) [inline db/db_impl/db_impl_open.cc]: `s/MinLogNumberToKeep()/min_log_number/`?

**Comment by ajkr** [comment]
Nice fix. The sequence of events makes sense to me.

File: db/db_impl/db_impl_open.cc [hunk 2]
```diff
@@ -588,12 +588,13 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& log_numbers,
   bool stop_replay_for_corruption = false;
   bool flushed = false;
   uint64_t corrupted_log_number = kMaxSequenceNumber;
+  uint64_t min_log_number = MinLogNumberToKeep();
   for (auto log_number : log_numbers) {
-    if (log_number < versions_->min_log_number_to_keep_2pc()) {
```
Comment (petermattis) [inline db/db_impl/db_impl_open.cc]: Isn't this a bug that is independent of `recycle_log_file_num`? When 2pc is disabled, `min_log_number_to_keep_2pc` is 0 which means all log files will be replayed, even if the manifest indicates a log file has already been fully processed. I'm not terribly familiar with the recovery code, so perhaps I'm missing something that increments `min_log_number_to_keep_2pc` when 2pc is disabled.

Comment (ajkr) [reply db/db_impl/db_impl_open.cc]: I don't think there's a problem with replaying a WAL that has already been fully processed. Each operation will be processed and ignored after checking the manifest, which tells us the operation's column family has already been flushed past this WAL. So reading unnecessary old WALs would do a lot of no-ops, in my understanding.

Comment (petermattis) [reply db/db_impl/db_impl_open.cc]: Ah, I was missing that `MemTableInserter` also checks that the recovering log number against the column family's log number.

**Pull request closed without merging**